### ISSUE TITLE
Broken test does not fail

### ISSUE THREAD

**Issue by colbec** [JuliaGPU/CUDAnative.jl#33]
Using:
```
Julia Version 0.6.0-dev.2464
Commit d233cf9* (2017-01-29 05:47 UTC)
Platform Info:
  OS: Linux (x86_64-suse-linux)
  CPU: Intel(R) Core(TM) i5-4460  CPU @ 3.20GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, haswell)
```
Pkg.test("CUDAdrv") passes but Pkg.test("CUDAnative") fails with two errors:
```
> Pkg.test("CUDAnative")
INFO: Testing CUDAnative
Test Summary:  | 
base interface | No tests
Test Summary:   | Pass  Total
code generation |   15     15
Test Summary: | Pass  Total
execution     |   10     10
Test Summary: | Pass  Total
device arrays |    2      2
printing: Error During Test
  Got an exception of type UndefVarError outside of a @test
  UndefVarError: ret not defined
  Stacktrace:
   [1] macro expansion at /home/colin/.julia/v0.6/CUDAnative/test/intrinsics.jl:34 [inlined]
   [2] macro expansion at ./test.jl:852 [inlined]
   [3] macro expansion at /home/colin/.julia/v0.6/CUDAnative/test/intrinsics.jl:0 [inlined]
   [4] macro expansion at ./test.jl:852 [inlined]
   [5] macro expansion at /home/colin/.julia/v0.6/CUDAnative/test/intrinsics.jl:0 [inlined]
   [6] macro expansion at ./test.jl:852 [inlined]
   [7] anonymous at ./<missing>:?
   [8] include_from_node1(::String) at ./loading.jl:539
   [9] include(::String) at ./sysimg.jl:14
```
also an error on CUDAdrv reported here since CUDAdrv tests pass, but it appears to be https://github.com/JuliaGPU/CUDAdrv.jl/issues/10 as reported by @timholy 

```
down: Error During Test
  Got an exception of type CUDAdrv.CuError outside of a @test
  No kernel image available/suitable for GPU (CUDA error #209, ERROR_NO_BINARY_FOR_GPU)
  ptxas application ptx input, line 67; error   : Instruction 'shfl' requires .target sm_30 or higher
  ptxas fatal   : Ptx assembly aborted due to errors
  Stacktrace:
```
CUDA

```
> cuda/NVIDIA_CUDA-8.0_Samples/bin/x86_64/linux/release/deviceQuery
cuda/NVIDIA_CUDA-8.0_Samples/bin/x86_64/linux/release/deviceQuery Starting...

 CUDA Device Query (Runtime API) version (CUDART static linking)

Detected 1 CUDA Capable device(s)

Device 0: "GeForce GT 610"
  CUDA Driver Version / Runtime Version          8.0 / 8.0
  CUDA Capability Major/Minor version number:    2.1
  Total amount of global memory:                 964 MBytes (1011023872 bytes)
  ( 1) Multiprocessors, ( 48) CUDA Cores/MP:     48 CUDA Cores
  GPU Max Clock rate:                            1620 MHz (1.62 GHz)
  Memory Clock rate:                             600 Mhz
  Memory Bus Width:                              64-bit
  L2 Cache Size:                                 65536 bytes
  Maximum Texture Dimension Size (x,y,z)         1D=(65536), 2D=(65536, 65535), 3D=(2048, 2048, 2048)
  Maximum Layered 1D Texture Size, (num) layers  1D=(16384), 2048 layers
  Maximum Layered 2D Texture Size, (num) layers  2D=(16384, 16384), 2048 layers
  Total amount of constant memory:               65536 bytes
  Total amount of shared memory per block:       49152 bytes
  Total number of registers available per block: 32768
  Warp size:                                     32
  Maximum number of threads per multiprocessor:  1536
  Maximum number of threads per block:           1024
  Max dimension size of a thread block (x,y,z): (1024, 1024, 64)
  Max dimension size of a grid size    (x,y,z): (65535, 65535, 65535)
  Maximum memory pitch:                          2147483647 bytes
  Texture alignment:                             512 bytes
  Concurrent copy and kernel execution:          Yes with 1 copy engine(s)
  Run time limit on kernels:                     No
  Integrated GPU sharing Host Memory:            No
  Support host page-locked memory mapping:       Yes
  Alignment requirement for Surfaces:            Yes
  Device has ECC support:                        Disabled
  Device supports Unified Addressing (UVA):      Yes
  Device PCI Domain ID / Bus ID / location ID:   0 / 1 / 0
  Compute Mode:
     < Default (multiple host threads can use ::cudaSetDevice() with device simultaneously) >

deviceQuery, CUDA Driver = CUDART, CUDA Driver Version = 8.0, CUDA Runtime Version = 8.0, NumDevs = 1, Device0 = GeForce GT 610
Result = PASS

```

**Comment by maleadt** [comment]
I think the CUDAnative test error has been fixed in f1f900e51ae1368722d5ac4685cc6a737457fd10, could you try the master branch?

That CUDAdrv error you mention isn't due to JuliaGPU/CUDAdrv.jl#10, because that would manifest itself when testing CUDAdrv. It seems like the shuffle tests fail because the selected target doesn't support `shfl`, and indeed your GeForce GT 610 only supports compute capability 2.1. I'll have a look at making those tests conditional.

**Comment by colbec** [comment]
Thankyou, with Pkg.checkout("CUDAnative") all tests now pass.

**Comment by colbec** [comment]
I just tried the test again without checking out the newest version and it passed again.

```
> Pkg.test("CUDAnative")
INFO: Testing CUDAnative
Test Summary:  | 
base interface | No tests
Test Summary:   | Pass  Total
code generation |   20     20
Test Summary: | Pass  Total
execution     |   10     10
Test Summary: | Pass  Total
device arrays |    2      2
INFO: CUDAnative tests passed

julia> Pkg.test("CUDAdrv")
INFO: Testing CUDAdrv
INFO: CUDAdrv tests passed
```
I checked my existing code for test/intrinsics.jl and indeed the new code is not there. 
So I dug into the code for the test which has now been isolated and ran it independently  against my card and got the error again.

Here is the test I used:
```
using CUDAdrv, CUDAnative
using Base.Test
dev = CuDevice(0)
ctx = CuContext(dev)
capability(dev) # produces v"2.1.0"
@testset "shuffle" begin
       n=14
       types = [Int32, Int64, Float32, Float64]
        
        @testset "down" begin
            @eval function kernel_shuffle_down{T}(d::CuDeviceArray{T}, n)
                t = threadIdx().x
                if t <= n
                    d[t] += shfl_down(d[t], n÷2)
                end
                return nothing
            end
        
            for T in types
                a = T[i for i in 1:n]
                d_a = CuArray(a)
        
                threads = nearest_warpsize(dev, n)
                @cuda (1, threads) kernel_shuffle_down(d_a, n)
        
                a[1:n÷2] += a[n÷2+1:end]
                @test a == Array(d_a)
            end
        end
        
end
```
which produces:
```
down: Error During Test
  Got an exception of type CUDAdrv.CuError outside of a @test
  No kernel image available/suitable for GPU (CUDA error #209, ERROR_NO_BINARY_FOR_GPU)
  ptxas application ptx input, line 49; error   : Instruction 'shfl' requires .target sm_30 or higher
  ptxas fatal   : Ptx assembly aborted due to errors
```
